### DISCUSSION TITLE
MPI checkpointing workaround

### DISCUSSION THREAD

**Discussion by ms609** [revbayes/revbayes#630] [General]
Some of my analyses are not meeting my ESS threshold after running for the maximum job time on my HPC server, so I've been thinking about how best to work around the current inability to restore an MPI MCMCMC run from checkpoints (#184).  

My current idea is to start a new MCMCMC analysis, using `.setValue` to set each variable to the values at the end of the existing analysis.
My thinking is that – assuming that PSRF analysis indicates that runs have converged – my initial analysis had reached the posterior, and thus starting from these values should avoid the need for a burnin phase and allow me to continue sampling from the posterior.  Adding the samples from the second analysis to the post-burnin samples from the first analysis would allow me to expand my sample size efficiently.

My concerns relate to my understanding that using `.setValue` would cause all runs and chains in my subsequent analysis to be initialized to a single draw.  This could:

  (i) invalidate further convergence diagnostics – perhaps not an issue if the runs have already converged;
  (ii) increase autocorrelation in early samples, as the hot chains would be initialized to sample the same region of parameter space as the cold chain.  This makes me feel that it might be necessary to "burn in" the second run before it was sampling the posterior as effectively as it would be if I'd restarted the run from a checkpoint.
  (iii) cause runs to be correlated, making it more difficult to compute the effective sample size when pooling all samples from both analyses.

The latter two concerns could be alleviated if there was a way to set separate values for each parameter in each chain in each run of an MCMCMC, but I don't think this can be accomplished at present.  They could perhaps be mitigated by including a short burnin in the second analysis, but I'm not sure how I'd determine when this burnin had been 'long enough' to be confident that the runs represented independent samples of the posterior.  I'd be interested to hear any thoughts!

**Comment by bjoelle** [comment]
I think the easiest way out may be to not assume anything about convergence of the first run, but to run the analysis from the new initial values as a separate thing without trying to pool with the first ? Using initial values from the first run should not be a problem because initial values do not change the posterior if convergence is reached. Hopefully the second run would converge fast enough that you get correct ESS from it alone...

**Reply by ms609** [reply]
I'm not sure how I would establish that the second analysis had converged, if all its constituent runs were initialized with the same parameter values.  I feel like it's almost necessary to test that the independent runs of the second analysis have _diverged_ enough that they are sampling independent regions of parameter space.  I don't know that ESS metrics would take into account the fact that at least iteration zero of each independent run in the second analysis would be correlated between runs.


(I've edited my original post to be more precise with my terminology: each time I launch RevBayes I'm starting a new **analysis**, which comprises `nruns` **runs**, each comprising `nchains` chains, one of which is the cold chain).

**Comment by davidcerny** [comment]
> The latter two concerns could be alleviated if there was a way to set separate values for each parameter in each chain in each run of an MCMCMC, but I don't think this can be accomplished at present.

Yes, I don't think that's possible, either. It should be possible in theory, though, because we generate checkpoint files not just for the cold chain, but also for each of the heated chains.

One ugly workaround could be to just write a new, separate script for each run. This would allow you to have run-specific starting values, though not chain-specific starting values. You might be able to get chain-specific values, and to eliminate the need for a bunch of `.setValue()` statements, if you are willing to execute these run-specific scripts using the single-core version of RevBayes. There is no obvious problem with restarting a single-run MC^3 analysis using the single-core version, so my guess would be that when `.initializeFromCheckpoint(...)` is called on an MC^3 object, it does in fact make use of the parameter values stored in the chain-specific checkpoint files. But I'd have to dig into the source code to make sure.

Of course, even if this actually worked, it would incur a significant performance hit. If you were previously using MPI to distribute _m_ runs of _n_ chains each over _m_ × _n_ cores, now you would only be using _m_ cores without MPI (you'd have to start _m_ single-core jobs in parallel, either manually or using something like a Slurm array).

But my guess is that it won't work. I previously tried something similar in despair – if I remember correctly, it wasn't an MC^3 analysis; I was just using MPI to parallelize a regular MCMC analysis with multiple runs – and found that the single-core version didn't like the checkpoint files generated by the MPI version. That would've been a while ago, though, so it's worth giving it another shot.

**Reply by ms609** [reply]
Hmm, separate scripts for separate runs is an interesting idea.  The obvious drawback is that one would lose the ability to check for convergence within RevBayes, but that is surmountable.  I guess we still need some way to seed the hot chains with something other than the cold chain values, whether this is by checkpointing (which means not using MPI), directly setting their values (not currently possible), or waiting (which probably means applying a sizeable burnin to follow-on analyses – arguably inefficient, but perhaps less so than running everything in serial?)

**Comment by bredelings** [comment]
On a slightly different note, _in theory_ it should be possible to save and restore applications using OS support.  See for example CRIU (https://criu.org).

In practice, this might work for containerized applications, but I'm not sure.  I see that singularity/apptainer claims to have support here: https://apptainer.org/docs/user/main/checkpoint.html.

Apparently they use DMTCP, which claims to be able to save and restore multithreaded and MPI applications that are started using SLURM: https://github.com/dmtcp/dmtcp

**Reply by ms609** [reply]
Thanks for the suggestion.  This would be a nice solution, but unfortunately the OS support available on my HPC only supports single-core jobs, not MPI applications.

**Comment by davidcerny** [comment]
Happy to report that between #553 and #646, checkpointing should be fully functional for MPI runs of MC^3 analyses (at least according to my quick test using @ms609's data from https://github.com/mk-nt/4310_by_ki), and workarounds should no longer be needed.

**Comment by davidcerny** [comment]
See #650: clearly, I've spoken too soon 🤦

**Reply by ms609** [reply]
Is #650 particular to resuming from an MPI checkpoint, or does this pre-existing issue that also affect resuming from a checkpoint in serial?

**Reply by davidcerny** [reply]
Unfortunately, it affects all MC^3 analyses, whether run using the single-core or the MPI version.
